src/firmware/app/primitives/direct_control_primitive.c

```diff
@@ -0,0 +1,147 @@
+                                        void* void_state_ptr, FirmwareWorld_t* world)
+{
+    DirectControlPrimitiveState_t* state = (DirectControlPrimitiveState_t*)void_state_ptr;
+    const FirmwareRobot_t* robot         = app_firmware_world_getRobot(world);
+    Chicker_t* chicker                   = app_firmware_robot_getChicker(robot);
+    Dribbler_t* dribbler                 = app_firmware_robot_getDribbler(robot);
+    Charger_t* charger                   = app_firmware_robot_getCharger(robot);
+
+    switch (prim_msg.which_wheel_control)
+    {
+        case DirectControlPrimitiveMsg_direct_per_wheel_control_tag:
+        {
+            DirectControlPrimitiveMsg_DirectPerWheelControlMsg control_msg =
+                prim_msg.wheel_control.direct_per_wheel_control;
+            state->direct_velocity = false;
+            // TODO (#1649): Fix passing rpm into an applyForce function
+            app_wheel_applyForce(app_firmware_robot_getFrontLeftWheel(robot),
+                                 control_msg.front_left_wheel_rpm);
+            app_wheel_applyForce(app_firmware_robot_getBackLeftWheel(robot),
+                                 control_msg.back_left_wheel_rpm);
+            app_wheel_applyForce(app_firmware_robot_getBackRightWheel(robot),
+                                 control_msg.back_right_wheel_rpm);
+            app_wheel_applyForce(app_firmware_robot_getFrontRightWheel(robot),
+                                 control_msg.front_right_wheel_rpm);
+            break;
+        }
+        case DirectControlPrimitiveMsg_direct_velocity_control_tag:
+        {
+            DirectControlPrimitiveMsg_DirectVelocityControlMsg control_msg =
+                prim_msg.wheel_control.direct_velocity_control;
+            state->direct_velocity          = true;
+            state->direct_target_velocity_x = control_msg.velocity.x_component_meters;
+            state->direct_target_velocity_y = control_msg.velocity.y_component_meters;
+            state->direct_target_velocity_angular =
+                control_msg.angular_velocity.radians_per_second;
+            break;
+        }
+        default:
+        {
+            // Do nothing
+            state->direct_velocity = false;
+            assert(false);
+        }
+    }
+
+    switch (prim_msg.charge_mode)
+    {
+        case DirectControlPrimitiveMsg_ChargeMode_CHARGE:
+        {
+            app_charger_charge_capacitor(charger);
+            break;
+        }
+        case DirectControlPrimitiveMsg_ChargeMode_FLOAT:
+        {
+            app_charger_float_capacitor(charger);
+            break;
+        }
+        case DirectControlPrimitiveMsg_ChargeMode_DISCHARGE:
+        {
+            app_charger_float_capacitor(charger);
+            app_chicker_disableAutochip(chicker);
+            app_chicker_disableAutokick(chicker);
+            break;
+        }
+        default:
+        {
+            // discharge
+            app_charger_float_capacitor(charger);
```

Comment:
should discharge capacitor :zap:

Reply:
Does
```
    app_chicker_disableAutochip(chicker);
    app_chicker_disableAutokick(chicker);
```
discharge the capacitors? 
firmware/boards/legacy_robot_stm32f4/io/charger.c doesn't have a discharge function...